File: src/main/java/me/kuku/yuq/controller/MiHoYoController.java
package me.kuku.yuq.controller;

import com.IceCreamQAQ.Yu.annotation.Action;
import com.IceCreamQAQ.Yu.annotation.Before;
import com.IceCreamQAQ.Yu.annotation.Config;
import com.IceCreamQAQ.Yu.annotation.Synonym;
import com.icecreamqaq.yuq.FunKt;
import com.icecreamqaq.yuq.annotation.GroupController;
import com.icecreamqaq.yuq.annotation.PrivateController;
import com.icecreamqaq.yuq.annotation.QMsg;
import me.kuku.yuq.entity.MiHoYoEntity;
import me.kuku.yuq.logic.MiHoYoLogic;
import me.kuku.yuq.pojo.Result;
import me.kuku.yuq.service.MiHoYoService;

import javax.inject.Inject;
import java.io.IOException;

@GroupController
public class MiHoYoController {
	@Inject
	private MiHoYoService miHoYoService;
	@Inject
	private MiHoYoLogic miHoYoLogic;

	@Before
	public MiHoYoEntity before(long qq){
		MiHoYoEntity miHoYoEntity = miHoYoService.findByQQ(qq);
		if (miHoYoEntity != null) return miHoYoEntity;
		else throw FunKt.getMif().at(qq).plus("您还没有绑定米哈游账号，请私聊机器人<mihoyo 账号 密码>进行绑定").toThrowable();
	}

	@Action("genshin")
	public String genShinUserInfo(MiHoYoEntity miHoYoEntity) throws IOException {
		return miHoYoLogic.genShinUserInfo(miHoYoEntity, null);
	}

	@PrivateController
	@GroupController
	public static class MiHoYoLoginController {
		@Inject
		private MiHoYoLogic miHoYoLogic;
		@Inject
		private MiHoYoService miHoYoService;
		@Config("YuQ.Mirai.bot.master")
		private String master;

		@Action("mihoyo {account} {password}")
		@Synonym({"米忽悠 {account} {password}", "米哈游 {account} {password}"})
		public String bindMiHoYo(String account, String password, long qq) throws IOException {
			Result<MiHoYoEntity> result = miHoYoLogic.login(account, password);
			if (result.isFailure()) return "绑定失败，" + result.getMessage();
			MiHoYoEntity newMiHoYoEntity = result.getData();
			MiHoYoEntity miHoYoEntity = miHoYoService.findByQQ(qq);
			if (miHoYoEntity == null) miHoYoEntity = new MiHoYoEntity(qq);
			miHoYoEntity.setAccount(account);
			miHoYoEntity.setPassword(password);
			miHoYoEntity.setCookie(newMiHoYoEntity.getCookie());
			miHoYoEntity.setAccountId(newMiHoYoEntity.getAccountId());
			miHoYoEntity.setTicket(newMiHoYoEntity.getTicket());
			miHoYoEntity.setCookieToken(newMiHoYoEntity.getCookieToken());
			miHoYoService.save(miHoYoEntity);
			return "绑定米忽悠成功！！";
		}


		@Action("genshin {id}")
		@QMsg(at = true, atNewLine = true)
		public String queryGenShinUserInfo(long id) throws IOException {
			MiHoYoEntity miHoYoEntity = miHoYoService.findByQQ(Long.parseLong(master));
			if (miHoYoEntity == null) return "无法查询，请联系机器人主人绑定原神账号（查询信息需要cookie）";
			return miHoYoLogic.genShinUserInfo(miHoYoEntity, id);
		}
	}
}

File: src/main/kotlin/me/kuku/yuq/controller/MiHoYoController.kt
package me.kuku.yuq.controller

import com.IceCreamQAQ.Yu.annotation.Action
import com.IceCreamQAQ.Yu.annotation.Before
import com.icecreamqaq.yuq.annotation.GroupController
import com.icecreamqaq.yuq.annotation.PrivateController
import com.icecreamqaq.yuq.controller.BotActionContext
import com.icecreamqaq.yuq.controller.ContextSession
import com.icecreamqaq.yuq.message.Message.Companion.firstString
import com.icecreamqaq.yuq.mif
import me.kuku.yuq.entity.MiHoYoEntity
import me.kuku.yuq.entity.MiHoYoService
import me.kuku.yuq.entity.QqEntity
import me.kuku.yuq.logic.MiHoYoLogic
import org.springframework.stereotype.Component

@GroupController
@PrivateController
@Component
class MiHoYoController (
    private val miHoYoService: MiHoYoService
) {

    @Action("米哈游登录")
    suspend fun login(session: ContextSession, qqEntity: QqEntity, context: BotActionContext): String {
        context.source.sendMessage("请发送手机号")
        val phone = session.waitNextMessage().firstString()
        context.source.sendMessage("请发送密码")
        val password = session.waitNextMessage().firstString()
        val result = MiHoYoLogic.login(phone, password)
        return if (result.success()) {
            val miHoYoEntity = result.data()
            val newEntity = miHoYoService.findByQqEntity(qqEntity) ?: MiHoYoEntity().also {
                it.qqEntity = qqEntity
            }
            newEntity.cookie = miHoYoEntity.cookie
            miHoYoService.save(newEntity)
            "绑定米哈游账号成功"
        } else result.message
    }

    @Before(except = ["login"])
    fun before(qqEntity: QqEntity): MiHoYoEntity {
        return miHoYoService.findByQqEntity(qqEntity) ?: throw mif.at(qqEntity.qq).plus("您没有绑定米哈游账号，操作失败").toThrowable()
    }

    @Action("原神签到")
    suspend fun sign(miHoYoEntity: MiHoYoEntity): String {
        val result = MiHoYoLogic.sign(miHoYoEntity)
        return if (result.success()) "原神签到成功"
        else "原神签到失败，${result.message}"
    }

    @Action("原神自动签到 {status}")
    fun open(status: Boolean, miHoYoEntity: MiHoYoEntity): String {
        miHoYoEntity.config.sign = status.toStatus()
        miHoYoService.save(miHoYoEntity)
        return "原神自动签到${if (status) "开启" else "关闭"}成功"
    }



}